include/aws/s3/s3_client.h

```diff
@@ -290,6 +306,32 @@ struct aws_s3_meta_request_options {
      * from the buffer and compare them them to previously uploaded part checksums.
      */
     const struct aws_byte_cursor *resume_token;
+
+    /**
+     * Optional.
+     * Proxy configuration for http connection.
+     */
+    struct aws_http_proxy_options *proxy_options;
+
+    /*
+     * Optional.
+     * Configuration for using proxy from environment variable.
+     * Defaults to true
+     * Only works when proxy_options is not set.
+     */
+    struct proxy_env_var_settings *proxy_ev_settings;
+
+    /**
+     * Optional.
+     * If set to 0, default value is used.
+     */
+    uint32_t connect_timeout_ms;
+
+    /*
+     * Optional.
+     * Set keepalive true to periodically transmit messages for detecting a disconnected peer.
+     */
+    struct aws_s3_tcp_keep_alive_options tcp_keep_alive_options;
```

Comment:
it's weird that we're adding these new options to `aws_s3_meta_request_options`

These options are used when creating an `aws_s3_endpoint` (which contains an `aws_http_connection_manager`). But then we store the `aws_s3_endpoint` in a hashtable and reuse them for later meta-requests. So if there are 2 meta-requests using the same endpoint, then the 2nd meta-request is going to use the `aws_s3_endpoint` that was created by the 1st meta-request. If the 1st meta-request didn't set any of these new options, but the 2nd meta-request did, then these new options will end up getting totally ignored.

If we keep these new options in `aws_s3_meta_request_options`, we need to be sure they're honored. Like, don't just look up pre-existing `aws_s3_endpoint` by URI, instead look up endpoints based on ALL these settings.

OR we go with another mechanism for configuring specific endpoints. Like, when creating the client, the user can pass in per-endpoint settings. Later on, when a meta-request is made, and we realize we need to create an `aws_s3_endpoint`, we have the settings to use for it

I don't know what level of configuration people are going to want though. Like, do they want to specify defaults? Like use these proxy options for ALL URI by default?  Or is there any chance they'd only want to use the extra settings for 1 request to a URI, but not other requests to that same URI?

Reply:
In the original two issues, customers are expecting to set these properties at the client level. S3 client in CPP (not the CRT-S3 client) also allowed customers to set these properties at a client level instead of an endpoint. I am not sure about how the other SDKs are handling it.
 
**Potential Solution 1 (Expose these options in the Client Config):**
I think we can also do something similar to S3 CPP client and move these properties to S3ClientConfig, which will resolve the issue of caching meta requests.

Issues:
- Customers will have to create different clients for different options.

**Potential Solution 2 (Expose these options per request basis):**
Current solution, customers will be able to configure different options for each endpoint. 

Issues:
- Caching (will need to discuss and decide on a solution)
- Customers will have to provide these properties for each meta request which might not be ideal if they don't change that frequently on per endpoint basis.

I am not sure either how frequently the customers will want to configure these properties. I am more inclined towards solution 1 as it is simple and matches the CPP S3 Client, however, this mainly depends on how frequently these options change.

Reply:
I am also inclined towards option 1 over option 2. It seems much more natural to configure the client, instead of each individual request.

We could ALSO allow per-endpoint configuration, by letting the user submit some kind of per-endpoint-configuration list. Then if a request uses that particular endpoint, it will use that configuration.